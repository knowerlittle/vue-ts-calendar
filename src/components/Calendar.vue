<template>
  <div :class="$style.calendar">
    <div
      v-if="isInfiniteScroll"
      :class="$style['calendar--infinite']"
    >
      <DayOfWeekNames
        :dayNames="dayNames"
        :sticky="true"
      />
      <Month
        v-for="(_, index) in monthsTillMaxDate"
        :key="index"
        :dayNames="dayNames"
        :monthNames="monthNames"
        :monthOffset="monthOffset + index"
        :showDayNames="false"
        :showPreviousMonth="false"
        :showNextMonth="false"
        :monthAlignRight="true"
        :currentSelectionRange="currentSelectionRange"
        :maximumSelectionRange="maximumSelectionRange"
        :isInfinite="isInfiniteScroll"
        @selectDate="selectDate"
        @changeMonth="updateMonth"
      />
    </div>

    <div
      v-else
      :class="{ [$style['calendar__flex']]: isDualView }"
    >
      <Month
        :dayNames="dayNames"
        :monthNames="monthNames"
        :offsetAmount="monthOffsetAmount"
        :monthOffset="monthOffset"
        :showPreviousMonth="true"
        :showNextMonth="isSingleView"
        :currentSelectionRange="currentSelectionRange"
        :maximumSelectionRange="maximumSelectionRange"
        @selectDate="selectDate"
        @changeMonth="updateMonth"
      />

      <template v-if="isDualView">
        <div :class="$style['calendar__divider']" />

        <Month
          :dayNames="dayNames"
          :monthNames="monthNames"
          :offsetAmount="2"
          :monthOffset="monthOffset + 1"
          :showPreviousMonth="false"
          :showNextMonth="true"
          :currentSelectionRange="currentSelectionRange"
          :maximumSelectionRange="maximumSelectionRange"
          @selectDate="selectDate"
          @changeMonth="updateMonth"
        />
      </template>
    </div>
  </div>
</template>

<script lang="ts">
import { DateTime, Interval } from 'luxon'
import { Component, Vue, Prop, Provide, Watch } from 'vue-property-decorator'

import Month from '@/components/Month.vue'
import DayOfWeekNames from '@/components/DayOfWeekNames.vue'
import CalendarDate from '@/services/CalendarDate'

import { ViewType, monthNamesFallback, dayNamesFallback } from '@/helpers'

const currentDate = DateTime.local()

  @Component({
    name: 'Calendar',
    components: {
      Month,
      DayOfWeekNames
    }
  })
export default class Calendar extends Vue {
    /**
     * Provides current date to all child components
     * @private
     */
    @Provide() currentDate = currentDate

    /**
     * Name of months
     */
    @Prop({
      default: () => monthNamesFallback
    })
    readonly monthNames!: string[]

    /**
     * Name of days
     */
    @Prop({
      default: () => dayNamesFallback
    })
    readonly dayNames!: string[]

    /**
     * Calendar display
     */
    @Prop({
      required: true,
      validator: value => Object.values(ViewType).includes(value)
    })
    readonly viewType!: string

    /**
     * Single or Range selection
     */
    @Prop({ default: true })
    readonly isRangeSelection!: boolean

    /**
     * Maximum Date for calendar
     */
    @Prop({ required: true })
    readonly maxDate!: string

    /**
     * Format date is emitted
     */
    @Prop({ default: '' })
    readonly dateFormat!: string

    $style: any

    startDate: DateTime | null = null
    endDate: DateTime | null = null
    monthOffset = 0

    get isSingleView () {
      return this.viewType === ViewType.Single
    }

    get isDualView () {
      return this.viewType === ViewType.Dual
    }

    get isInfiniteScroll () {
      return this.viewType === ViewType.Infinite
    }

    get monthOffsetAmount () {
      return this.isSingleView
        ? 1
        : 2
    }

    get maximumDate () {
      return DateTime.fromISO(this.maxDate)
    }

    get currentSelectionRange () {
      if (this.startDate && this.endDate) {
        return Interval.fromDateTimes(this.startDate, this.endDate)
      }
      return null
    }

    // Creates the range that is selectable
    get maximumSelectionRange () {
      return Interval.fromDateTimes(
        currentDate.plus({ days: -1 }),
        this.maximumDate.plus({ days: 1 })
      )
    }

    get monthsTillMaxDate () {
      return Math.ceil(this.maximumDate.diff(currentDate, 'months').months)
    }

    @Watch('endDate')
    onEndDateChange () {
      this.emitDate()
    }

    @Watch('viewType')
    onViewTypeChange () {
      this.monthOffset = 0
    }

    /**
     * Emits selected date
     * @type {Object}
     */
    private emitDate () {
      if (this.startDate && this.endDate) {
        const selectedDates = {
          fromDate: this.dateFormat
            ? this.startDate.toFormat(this.dateFormat)
            : this.startDate.toISO(),
          toDate: this.dateFormat
            ? this.endDate.toFormat(this.dateFormat)
            : this.endDate.toISO(),
          label: this.getSimpleDate()
        }

        this.$emit('dateSelection', selectedDates)
      }
    }

    getSimpleDate () {
      if (this.startDate && this.endDate) {
        const startDate = `${this.startDate.toFormat('dd LLL')}`
        const endDate = `${this.endDate.toFormat('dd LLL yyyy')}`

        return this.startDate.hasSame(this.endDate, 'day')
          ? endDate
          : `${startDate} - ${endDate}`
      }
    }

    selectDate (date: CalendarDate) {
      // User hasn't selected any dates
      // or it's a single date selection
      // if the date picked is before the range then set both dates as the new date
      // if both dates are selected in a range then reset date selection process
      if (this.startDate === null && this.endDate === null ||
        !this.isRangeSelection ||
        this.currentSelectionRange && this.currentSelectionRange.isAfter(date.luxon) ||
        this.startDate && this.endDate && !this.startDate.hasSame(this.endDate, 'day')
      ) {
        this.startDate = date.luxon
      }
      this.endDate = date.luxon
    }

    updateMonth (value: number) {
      this.monthOffset = this.monthOffset + value
    }
}
</script>

<style lang="scss" module>
  $infinite-view-margin-bottom: 50px;

  .calendar {
    &__flex {
      display: flex;
    }

    &__divider {
      height: 100%;
      border-left: 8px solid white;
      border-right: 4px solid white;
    }

    &--infinite {
      margin-bottom: $infinite-view-margin-bottom;
    }
  }
</style>
