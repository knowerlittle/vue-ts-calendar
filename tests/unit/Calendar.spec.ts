import { shallowMount } from '@vue/test-utils'
import { DateTime } from 'luxon'
import Calendar from '@/components/Calendar.vue'

// TODO: extract out to a separate __mocks__ folders; couldn't figure out how
jest.mock('luxon', () => {
  const mockLuxon = {
    toISO: jest.fn().mockReturnValue('22 2020'),
    toFormat: jest.fn().mockReturnValue('22 2020'),
    hasSame: jest.fn().mockReturnValue(true),
    diff: jest.fn().mockReturnValue({ months: 5 }),
    plus: jest.fn().mockReturnValue({
      local: jest.fn(),
      fromISO: jest.fn(),
      month: 2,
      year: 2020,
      plus: jest.fn()
    })
  }

  return {
    DateTime: {
      local: jest.fn().mockReturnValue(mockLuxon),
      fromISO: jest.fn().mockReturnValue(mockLuxon)
    },
    Interval: {
      hasSame: jest.fn(),
      contains: jest.fn(),
      plus: jest.fn(),
      fromDateTimes: jest.fn().mockReturnValue({
        fromDateTimes: jest.fn(),
        hasSame: jest.fn(),
        contains: jest.fn(),
        plus: jest.fn(),
        isAfter: jest.fn().mockReturnValue(true)
      })
    }
  }
})

describe('Calendar', () => {
  const factory = (opts = {}) => shallowMount(Calendar, {
    ...opts
  })

  describe('matches snapshot', () => {
    it('when it is single view', () => {
      const options = {
        propsData: {
          viewType: 'single',
          maxDate: '2020/12/20'
        }
      }

      const wrapper = factory(options)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  it('when it is dual view', () => {
    const options = {
      propsData: {
        viewType: 'dual',
        maxDate: '2020/12/20'
      }
    }

    const wrapper = factory(options)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('when it is infinite view', () => {
    const options = {
      propsData: {
        viewType: 'infinite',
        maxDate: '2020/12/20'
      }
    }

    const wrapper = factory(options)
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('events', () => {
    it('emits the correct object when date selection range is false', async () => {
      const options = {
        propsData: {
          viewType: 'single',
          maxDate: '2020/12/20',
          isRangeSelection: false
        }
      }
      const wrapper = factory(options)
      const dateMock = {
        luxon: DateTime.local()
      }

      wrapper.findComponent({ name: 'Month' }).vm.$emit('selectDate', dateMock)
      await wrapper.vm.$nextTick()

      // @ts-ignore
      expect(wrapper.emitted('dateSelection')[0][0]).toEqual({
        fromDate: '22 2020',
        toDate: '22 2020',
        label: '22 2020'
      })
    })
  })

  it('emits the correct object when date selection range is true', async () => {
    const options = {
      propsData: {
        viewType: 'single',
        maxDate: '2020/12/20',
        isRangeSelection: true
      }
    }
    const wrapper = factory(options)
    const dateMock1 = {
      luxon: DateTime.local()
    }
    const dateMock2 = {
      luxon: DateTime.local()
    }

    dateMock1.luxon.hasSame = jest.fn()
      .mockReturnValueOnce(false)
      .mockReturnValueOnce(true)

    wrapper.findComponent({ name: 'Month' }).vm.$emit('selectDate', dateMock1)
    wrapper.findComponent({ name: 'Month' }).vm.$emit('selectDate', dateMock2)

    await wrapper.vm.$nextTick()

    // @ts-ignore
    expect(wrapper.emitted('dateSelection')[0][0]).toEqual({
      fromDate: '22 2020',
      toDate: '22 2020',
      label: '22 2020 - 22 2020'
    })
  })

  it('changes the month offset by the emitted offset', () => {
    const options = {
      propsData: {
        viewType: 'single',
        maxDate: '2020/12/20'
      }
    }

    const wrapper = factory(options)

    wrapper.findComponent({ name: 'Month' }).vm.$emit('changeMonth', 1)

    expect(wrapper.vm.$data.monthOffset).toEqual(1)
  })

  it('resets month offset when viewType changes', async () => {
    const options = {
      propsData: {
        viewType: 'single',
        maxDate: '2020/12/20'
      }
    }
    const wrapper = factory(options)
    wrapper.setData({ monthOffset: 5 })
    // @ts-ignore
    expect(wrapper.vm.monthOffset).toEqual(5)

    wrapper.setProps({ viewType: 'infinite' })
    await wrapper.vm.$nextTick()

    // @ts-ignore
    expect(wrapper.vm.monthOffset).toEqual(0)
  })
})
