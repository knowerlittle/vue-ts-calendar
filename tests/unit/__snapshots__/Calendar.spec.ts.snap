// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar matches snapshot when it is single view 1`] = `
<div>
  <div class="">
    <month-stub showdaynames="true" maximumselectionrange="[object Object]" showpreviousmonth="true" shownextmonth="true" offsetamount="1" monthoffset="0" monthnames="January,February,March,April,May,June,July,August,September,October,November,December" daynames="Mon,Tues,Wed,Thu,Fri,Sat,Sun"></month-stub>
    <!---->
  </div>
</div>
`;

exports[`Calendar when it is dual view 1`] = `
<div>
  <div class="calendar__flex">
    <month-stub showdaynames="true" maximumselectionrange="[object Object]" showpreviousmonth="true" offsetamount="2" monthoffset="0" monthnames="January,February,March,April,May,June,July,August,September,October,November,December" daynames="Mon,Tues,Wed,Thu,Fri,Sat,Sun"></month-stub>
    <div class="calendar__divider"></div>
    <month-stub showdaynames="true" maximumselectionrange="[object Object]" shownextmonth="true" offsetamount="2" monthoffset="1" monthnames="January,February,March,April,May,June,July,August,September,October,November,December" daynames="Mon,Tues,Wed,Thu,Fri,Sat,Sun"></month-stub>
  </div>
</div>
`;

exports[`Calendar when it is infinite view 1`] = `
<div>
  <div class="calendar--infinite">
    <dayofweeknames-stub daynames="Mon,Tues,Wed,Thu,Fri,Sat,Sun" sticky="true"></dayofweeknames-stub>
    <month-stub maximumselectionrange="[object Object]" offsetamount="1" monthoffset="0" monthalignright="true" monthnames="January,February,March,April,May,June,July,August,September,October,November,December" daynames="Mon,Tues,Wed,Thu,Fri,Sat,Sun" isinfinite="true"></month-stub>
    <month-stub maximumselectionrange="[object Object]" offsetamount="1" monthoffset="1" monthalignright="true" monthnames="January,February,March,April,May,June,July,August,September,October,November,December" daynames="Mon,Tues,Wed,Thu,Fri,Sat,Sun" isinfinite="true"></month-stub>
    <month-stub maximumselectionrange="[object Object]" offsetamount="1" monthoffset="2" monthalignright="true" monthnames="January,February,March,April,May,June,July,August,September,October,November,December" daynames="Mon,Tues,Wed,Thu,Fri,Sat,Sun" isinfinite="true"></month-stub>
    <month-stub maximumselectionrange="[object Object]" offsetamount="1" monthoffset="3" monthalignright="true" monthnames="January,February,March,April,May,June,July,August,September,October,November,December" daynames="Mon,Tues,Wed,Thu,Fri,Sat,Sun" isinfinite="true"></month-stub>
    <month-stub maximumselectionrange="[object Object]" offsetamount="1" monthoffset="4" monthalignright="true" monthnames="January,February,March,April,May,June,July,August,September,October,November,December" daynames="Mon,Tues,Wed,Thu,Fri,Sat,Sun" isinfinite="true"></month-stub>
    <month-stub maximumselectionrange="[object Object]" offsetamount="1" monthoffset="5" monthalignright="true" monthnames="January,February,March,April,May,June,July,August,September,October,November,December" daynames="Mon,Tues,Wed,Thu,Fri,Sat,Sun" isinfinite="true"></month-stub>
  </div>
</div>
`;
